File: parsers/python/mdata.py
# Music Metadata • copyright blasieuwu, 2026
# a better way of getting track metadata.
from pathlib import Path
from typing import Any
import sys, re

REQUIRED = {
    "title",
    "artist",
    "album",
    "explicit"
}

def validate_mdata(file_name: str, silent: bool = False, yields: bool = False, track_identity=None) -> dict | None:
    directory = Path(__file__).parent.resolve()
    file_path = Path(directory / file_name).resolve()
    file_ext = Path(file_name).suffix

    if not file_path.exists():
        if not silent:
            print(f"\x1b[1;31m[1/2] [-] file wasn't located.\x1b[0m")
        return

    if file_ext != ".mdata":
        if not silent:
            print(f"\x1b[1;33m[2/2] [!] '{file_name}' is not a .mdata file...\x1b[0m")
        return

    if not silent:
        print(f"\n\x1b[1;32m--- processing {file_name} ---\x1b[0m\n")
    data = {}
    errors = []
    warnings = []
    identified = False

    lines = file_path.read_text(encoding="utf-8").splitlines()

    for line_num, line in enumerate(lines, 1):
        line = line.strip()

        # discard empty lines
        if not line:
            continue

        # syntax check // square brackets
        if not (line.startswith("[") and line.endswith("]")) or ":" not in line:
            errors.append(f"[{file_name}:{line_num}] invalid syntax -> '{line}'")
            continue

        content = line[1:-1].strip()
        key, value = content.split(":", 1)
        key, value = key.strip().lower(), value.strip()

        # syntax check // key/value args
        if not key or not value:
            errors.append(f"[{file_name}:{line_num}] empty key/value in '{line}'")
            continue

        # metadata checks
        if key == "explicit":
            if value.lower() not in ("true", "false"):
                errors.append(f"[{file_name}:{line_num}] 'explicit' must be 'true' or 'false'.")
            else:
                data[key] = value.lower() == "true"

        elif key in ("search-links", "playback-links"):
            links = [url.strip() for url in value.split("|")]
            data[key] = links

        elif key in ("artist", "romanized-artist"):
            data[key] = [a.strip() for a in value.split(",")]

        elif key in ("title", "romanized-title", "album"):
            data[key] = value

        elif key == "lyrics":
            data[key] = value
            lrc_file = file_path.parent / value
            if not lrc_file.exists():
                warnings.append(f"[{file_name}] lyrics file '{value}' missing.")

        else:
            errors.append(f"[{file_name}:{line_num}] unidentified metadata found -> '{line}'")

    if track_identity and not identified:
        title = data.get("romanized-title") or data.get("title")
        artist = data.get("romanized-artist") or data.get("artist")
        if title and artist:
            track_identity(title, artist)
            identified = True
    
    missing = REQUIRED - set(data.keys())
    if missing:
        errors.append(f"[{file_name}] missing required parameter(s): {', '.join(missing)}")

    if errors:
        if not silent:
            print(f"\x1b[1;31m[-] status: FAILED\x1b[0m")
            for error in errors:
                print(f"\x1b[1;31m• [E] {error}\x1b[0m")
        return None
    if not silent:    
        print(f"\x1b[1;32m[+] status: PASSED\x1b[0m")
        for warn in warnings:
            print(f"\x1b[1;33m• [!] {warn}\x1b[0m")

        print("\n\x1b[1;36m[+] parsed metadata:\x1b[0m")
        print(data)
        
    return data

def _track_identify(file_name: str) -> Any | dict[str, str]:
    directory = Path(__file__).parent.resolve()
    file_path = Path(directory / file_name).resolve()
    file_ext = Path(file_name).suffix
    
    if not file_path.exists():
        return
    
    if file_ext != ".mdata":
        return
    
    track = {}
    
    lines = file_path.read_text(encoding="utf-8").splitlines()
    
    for line in lines:
        line = line.strip()
    
        # discard empty lines
        if not line:
            continue
    
        # syntax check // square brackets
        if not (line.startswith("[") and line.endswith("]")) or ":" not in line:
            continue
    
        content = line[1:-1].strip()
        key, value = content.split(":", 1)
        key, value = key.strip().lower(), value.strip()
    
        if key in ("artist", "romanized-artist"):
            track['artist'] = [a.strip() for a in value.split(",")]

        elif key in ("title", "romanized-title"):
            track['title'] = value

        elif key == "explicit":
            if value.lower() not in ("true", "false"):
                continue
            else:
                track['explicit'] = value.lower()
    
        else:
            continue

    return track

def multi_file_input():
    files = []
    print("\nenter file names (or hit Enter to finish)")
    while True:
        file = input("\x1b[1;38;2;12;59;5m[?] .mdata file to parse -> ")
        sys.stdout.write("\x1b[0m")
        sys.stdout.flush()

        if file == "":
            break

        files.append(file)

    for mdata_file in files:
        validate_mdata(mdata_file)

def single_file_input():
    file = input("\x1b[1;38;2;12;59;5m[?] .mdata file to parse -> ")
    sys.stdout.write("\x1b[0m")
    sys.stdout.flush()
    
    validate_mdata(file)


# for systems that have to programmically pass files without user input
def singlefile(file: str, silent: bool = False) -> dict | None:
    return validate_mdata(file, silent=silent)

def multifile(*files: str, silent: bool = False) -> list[dict | None]:
    return [validate_mdata(file, silent=silent) for file in files]

def singletrack_identity(file: str) -> None:
    identity = _track_identify(file)
    if not identity:
        return
    
    print(f"\x1b[1m[{file}] track info:\x1b[0m")
    print(f"- title: \x1b[1m{identity['title']}\x1b[0m")
    print(f"- artist(s): \x1b[1m{', '.join(identity['artist'])}\x1b[0m")
    print(f"- explicit: \x1b[1m{identity['explicit']}\x1b[0m")

def multitrack_identify(*files: str):
    for file in files:
        singletrack_identity(file)

def singlefile_autoname(file: str):
    directory = Path(__file__).parent.resolve()
    file_path = Path(directory / file).resolve()
    file_ext = Path(file).suffix
    identity = _track_identify(file)
    if not identity or 'title' not in identity or 'artist' not in identity:
        return

    title = identity['title']
    artists = ', '.join(identity['artist']) if isinstance(identity['artist'], list) else identity['artist']

    _temp = re.sub(r"[^\w\s]", "", title)
    filetitle = re.sub(r"\s+", "_", _temp.strip())

    _temp = re.sub(r"[^\w\s]", "", artists)
    fileartists = re.sub(r"\s+", "_", _temp.strip())

    fullfile = f"{filetitle.lower()}-{fileartists.lower()}.mdata"

    new_file = file_path.parent / fullfile
    file_path.replace(new_file)

def multifile_autoname(*files: str):
    for file in files:
        singlefile_autoname(file)

def directory_mdata(dir_path: str = ".", recursive: bool = False, silent: bool = False) -> list[dict | None]:
    target_dir = Path(__file__).parent.resolve() / dir_path
    
    if not target_dir.is_dir():
        if not silent:
            print(f"\x1b[1;31m[-] directory '{dir_path}' not found.\x1b[0m")
        return []

    # search recursively or just top-level
    pattern = "**/*.mdata" if recursive else "*.mdata"
    mdata_files = [f.name for f in target_dir.glob(pattern)]

    if not mdata_files and not silent:
        print(f"\x1b[1;33m[!] no .mdata files found in '{dir_path}'.\x1b[0m")
        return []

    return [validate_mdata(file, silent=silent) for file in mdata_files]


def directory_autoname(dir_path: str = ".", recursive: bool = False):
    target_dir = Path(__file__).parent.resolve() / dir_path

    if not target_dir.is_dir():
        print(f"\x1b[1;31m[-] directory '{dir_path}' not found.\x1b[0m")
        return

    pattern = "**/*.mdata" if recursive else "*.mdata"
    for file_path in target_dir.glob(pattern):
        # pass relative path from directory root
        singlefile_autoname(file_path.name)

def main():
    while True:
        print("\x1b[1mselect mode:\x1b[0m")
        print("1. single file")
        print("2. multiple files")
        print("3. entire folder")
        
        mode = input("\x1b[1;38;2;12;59;5m[?] option (1/2/3) -> \x1b[0m").strip()
        
        if mode == "1":
            single_file_input()
            break
        elif mode == "2":
            multi_file_input()
            break
        elif mode == "3":
            folder = input("\x1b[1;38;2;12;59;5m[?] folder path (leave empty for current dir) -> \x1b[0m").strip() or "."
            rec = input("\x1b[1;38;2;12;59;5m[?] include subfolders? (y/n) -> \x1b[0m").lower().startswith("y")
            directory_mdata(folder, recursive=rec)
            break
        else:
            print("\x1b[1;33m[!] invalid option.\x1b[0m\n")
    
if __name__ == "__main__":
    main()

File: parsers/typescript/mdata.ts
// Music Metadata • copyright blasieuwu, 2026
// a better way of getting track metadata.
import * as fs from 'fs';
import * as path from 'path';
import * as readline from 'readline';
import { fileURLToPath } from 'url';

export const REQUIRED: ReadonlySet<string> = new Set([
  'title',
  'artist',
  'album',
  'explicit'
]);

export interface MDataResult {
  title?: string;
  artist?: string[];
  album?: string;
  explicit?: boolean;
  'romanized-title'?: string;
  'romanized-artist'?: string[];
  'search-links'?: string[];
  'playback-links'?: string[];
  lyrics?: string;
  [key: string]: any;
}

export interface TrackIdentity {
  title?: string;
  artist?: string[];
  explicit?: string;
}

export function validateMdata(
  fileName: string,
  silent: boolean = false,
  yields: boolean = false,
  trackIdentity?: (title: string, artist: string[]) => void
): MDataResult | null {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const filePath = path.resolve(currentDir, fileName);
  const fileExt = path.extname(fileName);

  if (!fs.existsSync(filePath)) {
    if (!silent) {
      console.log('\x1b[1;31m[1/2] [-] file wasn\'t located.\x1b[0m');
    }
    return null;
  }

  if (fileExt !== '.mdata') {
    if (!silent) {
      console.log(`\x1b[1;33m[2/2] [!] '${fileName}' is not a .mdata file...\x1b[0m`);
    }
    return null;
  }

  if (!silent) {
    console.log(`\n\x1b[1;32m--- processing ${fileName} ---\x1b[0m\n`);
  }

  const data: MDataResult = {};
  const errors: string[] = [];
  const warnings: string[] = [];
  let identified = false;

  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split(/\r?\n/);

  lines.forEach((rawLine: string, index: number) => {
    const lineNum = index + 1;
    const line = rawLine.trim();

    // discard empty lines
    if (!line) return;

    // syntax check // square brackets
    if (!(line.startsWith('[') && line.endsWith(']')) || !line.includes(':')) {
      errors.push(`[${fileName}:${lineNum}] invalid syntax -> '${line}'`);
      return;
    }

    const innerContent = line.slice(1, -1).trim();
    const colonIdx = innerContent.indexOf(':');
    const key = innerContent.slice(0, colonIdx).trim().toLowerCase();
    const value = innerContent.slice(colonIdx + 1).trim();

    // syntax check // key/value args
    if (!key || !value) {
      errors.push(`[${fileName}:${lineNum}] empty key/value in '${line}'`);
      return;
    }

    // metadata checks
    if (key === 'explicit') {
      if (!['true', 'false'].includes(value.toLowerCase())) {
        errors.push(`[${fileName}:${lineNum}] 'explicit' must be 'true' or 'false'.`);
      } else {
        data[key] = value.toLowerCase() === 'true';
      }
    } else if (key === 'search-links' || key === 'playback-links') {
      data[key] = value.split('|').map(url => url.trim());
    } else if (key === 'artist' || key === 'romanized-artist') {
      data[key] = value.split(',').map(a => a.trim());
    } else if (key === 'title' || key === 'romanized-title' || key === 'album') {
      data[key] = value;
    } else if (key === 'lyrics') {
      data[key] = value;
      const lrcFile = path.resolve(path.dirname(filePath), value);
      if (!fs.existsSync(lrcFile)) {
        warnings.push(`[${fileName}] lyrics file '${value}' missing.`);
      }
    } else {
      errors.push(`[${fileName}:${lineNum}] unidentified metadata found -> '${line}'`);
    }
  });

  if (trackIdentity && !identified) {
    const title = data['romanized-title'] || data['title'];
    const artist = data['romanized-artist'] || data['artist'];
    if (title && artist) {
      trackIdentity(title, artist);
      identified = true;
    }
  }

  const keysPresent = new Set(Object.keys(data));
  const missing = Array.from(REQUIRED).filter(k => !keysPresent.has(k));

  if (missing.length > 0) {
    errors.push(`[${fileName}] missing required parameter(s): ${missing.join(', ')}`);
  }

  if (errors.length > 0) {
    if (!silent) {
      console.log('\x1b[1;31m[-] status: FAILED\x1b[0m');
      errors.forEach(err => console.log(`\x1b[1;31m• [E] ${err}\x1b[0m`));
    }
    return null;
  }

  if (!silent) {
    console.log('\x1b[1;32m[+] status: PASSED\x1b[0m');
    warnings.forEach(warn => console.log(`\x1b[1;33m• [!] ${warn}\x1b[0m`));

    console.log('\n\x1b[1;36m[+] parsed metadata:\x1b[0m');
    console.log(data);
  }

  return data;
}

export function _track_identify(fileName: string): TrackIdentity | null {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const filePath = path.resolve(currentDir, fileName);
  const fileExt = path.extname(fileName);

  if (!fs.existsSync(filePath) || fileExt !== '.mdata') {
    return null;
  }

  const track: TrackIdentity = {};
  const content = fs.readFileSync(filePath, 'utf-8');
  const lines = content.split(/\r?\n/);

  for (const rawLine of lines) {
    const line = rawLine.trim();

    if (!line || !(line.startsWith('[') && line.endsWith(']')) || !line.includes(':')) {
      continue;
    }

    const innerContent = line.slice(1, -1).trim();
    const colonIdx = innerContent.indexOf(':');
    const key = innerContent.slice(0, colonIdx).trim().toLowerCase();
    const value = innerContent.slice(colonIdx + 1).trim();

    if (key === 'artist' || key === 'romanized-artist') {
      track.artist = value.split(',').map((a: string) => a.trim());
    } else if (key === 'title' || key === 'romanized-title') {
      track.title = value;
    } else if (key === 'explicit') {
      if (!['true', 'false'].includes(value.toLowerCase())) {
        continue;
      }
      track.explicit = value.toLowerCase();
    }
  }

  return track;
}

export function prompt(query: string): Promise<string> {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  return new Promise(resolve => {
    rl.question(query, (ans: string) => {
      rl.close();
      resolve(ans);
    });
  });
}

export async function multi_file_input(): Promise<void> {
  const files: string[] = [];
  console.log('\nenter file names (or hit Enter to finish)');
  while (true) {
    const file = await prompt('\x1b[1;38;2;12;59;5m[?] .mdata file to parse -> \x1b[0m');
    if (file === '') {
      break;
    }
    files.push(file);
  }

  for (const mdata_file of files) {
    validateMdata(mdata_file);
  }
}

export async function single_file_input(): Promise<void> {
  const file = await prompt('\x1b[1;38;2;12;59;5m[?] .mdata file to parse -> \x1b[0m');
  validateMdata(file);
}

// for systems that have to programmically pass files without user input
export function singlefile(file: string, silent: boolean = false): MDataResult | null {
  return validateMdata(file, silent);
}

export function multifile(...files: string[]): (MDataResult | null)[] {
  return files.map(file => validateMdata(file));
}

export function singletrack_identity(file: string): void {
  const identity = _track_identify(file);
  if (!identity) return;

  console.log(`\x1b[1m[${file}] track info:\x1b[0m`);
  console.log(`- title: \x1b[1m${identity.title}\x1b[0m`);
  console.log(`- artist(s): \x1b[1m${identity.artist?.join(', ')}\x1b[0m`);
  console.log(`- explicit: \x1b[1m${identity.explicit}\x1b[0m`);
}

export function multitrack_identify(...files: string[]): void {
  for (const file of files) {
    singletrack_identity(file);
  }
}

export function singlefile_autoname(file: string): void {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const filePath = path.resolve(currentDir, file);
  const identity = _track_identify(file);

  if (!identity || !identity.title || !identity.artist) {
    return;
  }

  const title = identity.title;
  const artists = Array.isArray(identity.artist) ? identity.artist.join(', ') : identity.artist;

  const tempTitle = title.replace(/[^\w\s]/g, '');
  const filetitle = tempTitle.trim().replace(/\s+/g, '_');

  const tempArtists = artists.replace(/[^\w\s]/g, '');
  const fileartists = tempArtists.trim().replace(/\s+/g, '_');

  const fullfile = `${filetitle.toLowerCase()}-${fileartists.toLowerCase()}.mdata`;
  const newFile = path.resolve(path.dirname(filePath), fullfile);

  fs.renameSync(filePath, newFile);
}

export function multifile_autoname(...files: string[]): void {
  for (const file of files) {
    singlefile_autoname(file);
  }
}

function getMdataFiles(dirPath: string, recursive: boolean): string[] {
  let results: string[] = [];
  const entries = fs.readdirSync(dirPath, { withFileTypes: true });

  for (const entry of entries) {
    const fullPath = path.join(dirPath, entry.name);
    if (entry.isDirectory() && recursive) {
      results = results.concat(getMdataFiles(fullPath, recursive));
    } else if (entry.isFile() && entry.name.endsWith('.mdata')) {
      results.push(fullPath);
    }
  }
  return results;
}

export function directory_mdata(dirPath: string = '.', recursive: boolean = false, silent: boolean = false): (MDataResult | null)[] {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const targetDir = path.resolve(currentDir, dirPath);

  if (!fs.existsSync(targetDir) || !fs.statSync(targetDir).isDirectory()) {
    if (!silent) {
      console.log(`\x1b[1;31m[-] directory '${dirPath}' not found.\x1b[0m`);
    }
    return [];
  }

  const mdataFiles = getMdataFiles(targetDir, recursive).map(f => path.relative(currentDir, f));

  if (mdataFiles.length === 0 && !silent) {
    console.log(`\x1b[1;33m[!] no .mdata files found in '${dirPath}'.\x1b[0m`);
    return [];
  }

  return mdataFiles.map(file => validateMdata(file, silent));
}

export function directory_autoname(dirPath: string = '.', recursive: boolean = false): void {
  const currentDir = path.dirname(fileURLToPath(import.meta.url));
  const targetDir = path.resolve(currentDir, dirPath);

  if (!fs.existsSync(targetDir) || !fs.statSync(targetDir).isDirectory()) {
    console.log(`\x1b[1;31m[-] directory '${dirPath}' not found.\x1b[0m`);
    return;
  }

  const mdataFiles = getMdataFiles(targetDir, recursive).map(f => path.relative(currentDir, f));
  for (const filePath of mdataFiles) {
    singlefile_autoname(filePath);
  }
}

export async function main(): Promise<void> {
  while (true) {
    console.log('\x1b[1mselect mode:\x1b[0m');
    console.log('1. single file');
    console.log('2. multiple files');
    console.log('3. entire folder');

    const mode = (await prompt('\x1b[1;38;2;12;59;5m[?] option (1/2/3) -> \x1b[0m')).trim();

    if (mode === '1') {
      await single_file_input();
      break;
    } else if (mode === '2') {
      await multi_file_input();
      break;
    } else if (mode === '3') {
      const folderInput = await prompt('\x1b[1;38;2;12;59;5m[?] folder path (leave empty for current dir) -> \x1b[0m');
      const folder = folderInput.trim() || '.';
      const recInput = await prompt('\x1b[1;38;2;12;59;5m[?] include subfolders? (y/n) -> \x1b[0m');
      const rec = recInput.trim().toLowerCase().startsWith('y');
      directory_mdata(folder, rec);
      break;
    } else {
      console.log('\x1b[1;33m[!] invalid option.\x1b[0m\n');
    }
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}